qldpc/codes.py

```diff
@@ -675,6 +761,382 @@ def _minimize_weight_of_logical_op(
+        matrix = np.block(
+            [
+                [np.zeros_like(self.code_z.matrix), self.code_z.matrix],
+                [self.code_x.matrix, np.zeros_like(self.code_x.matrix)],
+            ]
+        ).reshape((self.num_checks, 2, -1))
+        return matrix
+
+    @property
+    def num_checks(self) -> int:
+        """Number of parity checks."""
+        return self.code_x.matrix.shape[0] + self.code_z.matrix.shape[0]
+
+    @property
+    def num_qubits(self) -> int:
+        """Number of data qubits in this code."""
+        return self.code_x.matrix.shape[1]
+
+    @property
+    def num_logical_qubits(self) -> int:
+        """Number of logical qubits encoded by this code."""
+        return self.code_x.dimension + self.code_z.dimension - self.num_qubits
+
+    """
+    Code tranformation methods — Conjugation and Shifting.
+
+    (i) Conjugation — A CSSCode can specify which data qubits should be hadamard-transformed
+    before/after syndrome extraction, thereby transforming the operators
+    that address a specified data qubit as (X,Y,Z) <--> (Z,Y,X).
+
+    (ii) Shifting — A CSSCode can also "shift" the Pauli operators on a qubit, moving vertically
+    along the following table:
+
+    ―――――――――――
+    | XY | YX |
+    ―――――――――――
+    | XZ | ZX |
+    ―――――――――――
+    | YZ | ZY |
+    ―――――――――――
+
+    Qubit shifts are specified by a dictionary mapping qubit_index --> shift_index, where
+    shift_index = +1, 0, and -1 mod 3 respectively refer to the top, middle, and bottom rows of the
+    table.
+
+    Physically, a shift of +1 and -1 respectively correspond to conjugating the Pauli operators
+    addressing a qubit by sqrt(X) and sqrt(Z) rotations.
+    """
+
+    # Given a parity check matrix, swap X and Z operators on the qubits to conjugate.
+    # TODO: Does this work for general stabilizer code (over F_2)? If so, move to QuditCode class?
+    def conjugate(cls, matrix: IntegerMatrix | None, conj: slice | Sequence[int]) -> IntegerMatrix:
+        matrix[:, :, conj] = np.roll(matrix[:, :, conj], 1, axis=1)
+        return matrix
+
+    def shift(cls, matrix: IntegerMatrix | None, shifts: dict[int, int]) -> IntegerMatrix:
+        # identify qubits to shift up or down along the table of Pauli pairs
+        shifts_up = tuple(qubit for qubit, shift in shifts.items() if shift % 3 == 1)
+        shifts_dn = tuple(qubit for qubit, shift in shifts.items() if shift % 3 == 2)
+        # For qubits shifting up, any stabilizer addressing a qubit with a Z should now also
+        # address that qubit with X, so copy Z to X.  Likewise for shifting down and X to Z.
+        matrix[:, Pauli.X.index, shifts_up] |= matrix[:, Pauli.Z.index, shifts_up]
+        matrix[:, Pauli.Z.index, shifts_dn] |= matrix[:, Pauli.X.index, shifts_dn]
+        return matrix
+
+    """Code Paramenter Computation methods
+
+        (i) Calculating distance, both exactly and upper bounds.
+
+        (ii) Obtaining Logical operators (remove?).
```

Comment:
Do not remove.  Logical operators are definitely useful for people working downstream with circuits and algorithms :)

Reply:
Gotcha!